### ISSUE TITLE
minion should be able to report to which nodegroups it belongs

### ISSUE THREAD

**Issue by ahammond** [saltstack/salt#33125]
### Description of Issue/Question
Nodegroups (aka, the poor man's CMDB) are awesome. Minions can already decide if they are a member or not of a given nodegroup for the purposes of targeting. It'd be very helpful to make this information visible as a grain:

```yaml
nodegroups:
  - foo
  - bar
```

**Comment by thatch45** [comment]
So you mean you want the nodegroups defined on the master to show up as grains on the minion if the minion is part of that nodegroup?
If that is the case then I would say that this sounds more like an ext_pillar, since the association data is on the master so logically it should be in the pillar

**Comment by cachedout** [comment]
Yes, I agree. We could probably find a way to inject this into the pillar. Then we could even use the pillar runner to filter and sort it if we wanted.

**Comment by thatch45** [comment]
Unless I am totally missing what @ahammond is saying, something that happens more often that I would generally like to admit :)

**Comment by ahammond** [comment]
@thatch45 I think we're speaking the same language here. My goal is to be able to know which nodegroups a minion belongs to while inside states. Then I can do things like

```jinja
{%- if 'my_role' in pillar.nodegroups %}
```

The alternative is to use a roles grain, but then I have to populate that, which is an extra step. @white-hat this will work for our use-case, right?

**Comment by thatch45** [comment]
yes, looks like an external pillar will work, since you would just need to add the nodegroups from opts. Or you could turn on the master config pillar, then the nodegroups would be under pillar:master:nodegroups

**Comment by ahammond** [comment]
So just something like:

def ext_pillar(minion_id, pillar):
    return __opts__['nodegroups']

**Comment by cachedout** [comment]
That's clever. I like it.

**Comment by damon-atkins** [comment]
Keep in mine that gains can not be trusted.  e.g. Break into a minion, changes its role grain, then get extra access or extra passwords etc, when the next high state runs.

The only way you can trust a minion to provide the role information is if its inside a minions signed certificate.  So I like your idea.

**Comment by ahammond** [comment]
Yeah, grains are a no-go. I tried this:

def ext_pillar(minion_id, pillar):
    return {'nodegroups': __opts__['nodegroups']}

But unfortunately, that just gives me all my nodegroups. What I want is the
resolved list of nodegroups of which the minion is a member.

**Comment by thatch45** [comment]
in the ext pillar you have all of the grains etc of the minion, you can instantiate the Matcher class and executes the nodegroup matcher for each nodegroup on the available minion data

**Comment by ahammond** [comment]
I have the following:

```python
from salt.minion import Matcher


def ext_pillar(minion_id, pillar, *args, **kwargs):
    nodegroups = []
    m = Matcher(__opts__)

    for k, v in __opts__['nodegroups'].iteritems():
        if m.nodegroup_match(minion_id, k):
            nodegroups.append(k)

    return {'nodegroups': nodegroups}
```

But it fails on the `m.nodegroup_match()` call. The code comment for Minion.nodegroup_match() doesn't show how to actually use it. Can I please get an example usage?

**Comment by ahammond** [comment]
This works. 

```python
from salt.minion import Matcher


def ext_pillar(minion_id, pillar, *args, **kwargs):
    l = getLogger('ext_pillar.nodegroups')
    nodegroups = []
    m = Matcher(__opts__)

    for k, v in __opts__['nodegroups'].iteritems():
        if m.nodegroup_match(k, __opts__['nodegroups']):
            nodegroups.append(k)

    return {'nodegroups': nodegroups}
```

@thatch45 How / where would I add it to submit it as a PR?

**Comment by thatch45** [comment]
glad you got it, sorry I was not able to help out more! just make a new module called nodegroup.py in salt/pillar/ and send a pr in on the develop branch

**Comment by rallytime** [comment]
Fixed by #33179
